### PR TITLE
Avoid allocations in resize and realloc

### LINKED ISSUE #4193: Unnecessary code in DualView::resize on host
Currently `DualView::resize` on the host takes a very convoluted path with multiple calls to `create_mirror_view`, see https://github.com/kokkos/kokkos/blob/master/containers/src/Kokkos_DualView.hpp#L912. I think this code should be replaced with something simpler like this to match the resize codepath for the device:

```
      ::Kokkos::resize(h_view, n0, n1, n2, n3, n4, n5, n6, n7);
      d_view = create_mirror_view(h_view);
```

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #4193.

`resize()` and `realloc` always create new objects even if the extents requested are the same as the container already has. In some way, this is a continuation of https://github.com/kokkos/kokkos/pull/916 that already does this optimization for `Kokkos::resize(Kokkos::View, ...)`.
However, I believe that we need to check `dynamic_rank` instead of `Rank` (or `rank`) since we only care about the dynamic extents anyway and comparing the static extents with the requested ones would always fail (if the static rank is greater than 0).
For `resize` we don't do anything if the extents already match but we call Kokkos::deep_copy for `realloc` to preserve the current behavior. This should of course change when we add the `AllocateWithoutInitializing` overloads.

### CODE REVIEW

**Comment by masterleinad** [comment]
For
```
#include <Kokkos_Core.hpp>
#include <Kokkos_DualView.hpp>
#include <Kokkos_DynRankView.hpp>
#include <cstdio>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);

  {
    Kokkos::View<int****[1][2][3][4]> bla("bla", 5,6,7,8);

    Kokkos::Profiling::pushRegion("Resize same view");
    Kokkos::resize(bla, 5,6,7,8);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Resize other view");
    Kokkos::resize(bla, 5,6,7,9);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Realloc same view");
    Kokkos::realloc(bla, 5,6,7,9);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Realloc other view");
    Kokkos::realloc(bla, 5,6,7,8);
    Kokkos::Profiling::popRegion();
  }

  {
    Kokkos::DualView<int****[1][2][3][4]> bla("bla", 5,6,7,8);

    Kokkos::Profiling::pushRegion("Resize same dualview");
    Kokkos::resize(bla, 5,6,7,8);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Resize other dualview");
    Kokkos::resize(bla, 5,6,7,9);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Realloc same dualview");
    Kokkos::realloc(bla, 5,6,7,9);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Realloc other dualview");
    Kokkos::realloc(bla, 5,6,7,8);
    Kokkos::Profiling::popRegion();
  }

  {
    Kokkos::DynRankView<int> bla("bla", 5,6,7,8);

    Kokkos::Profiling::pushRegion("Resize same dynrankview");
    Kokkos::resize(bla, 5,6,7,8);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Resize other dynrankvieww");
    Kokkos::resize(bla, 5,6,7,9);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Realloc same dynrankview");
    Kokkos::realloc(bla, 5,6,7,9);
    Kokkos::Profiling::popRegion();

    Kokkos::Profiling::pushRegion("Realloc other dynrankview");
    Kokkos::realloc(bla, 5,6,7,8);
    Kokkos::Profiling::popRegion();
  }

  Kokkos::finalize();
}
```
I am seeing
```
KokkosP: Example Library Initialized (sequence is 0, version: 20210623)
KokkosP: Allocate<Host> name: bla pointer: 0x7f1d284cd040 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 0
KokkosP: Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 0 is completed.
KokkosP: Entering profiling region: Resize same view
KokkosP: Allocate<Host> name: bla pointer: 0x7f1d28497040 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 1
KokkosP: Resize same view
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 1 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 2
KokkosP: Resize same view
KokkosP:   Kokkos::ViewCopy-8D
KokkosP: Execution of kernel 2 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x7f1d284cd040 size: 161280
KokkosP: Exiting profiling region: Resize same view
KokkosP: Entering profiling region: Resize other view
KokkosP: Allocate<Host> name: bla pointer: 0x7f1d284c8040 size: 181440
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 3
KokkosP: Resize other view
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 3 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 4
KokkosP: Resize other view
KokkosP:   Kokkos::ViewCopy-8D
KokkosP: Execution of kernel 4 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x7f1d28497040 size: 161280
KokkosP: Exiting profiling region: Resize other view
KokkosP: Entering profiling region: Realloc same view
KokkosP: Deallocate<Host> name: bla pointer: 0x7f1d284c8040 size: 181440
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f881ec0 size: 181440
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 5
KokkosP: Realloc same view
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 5 is completed.
KokkosP: Exiting profiling region: Realloc same view
KokkosP: Entering profiling region: Realloc other view
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f881ec0 size: 181440
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f881ec0 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 6
KokkosP: Realloc other view
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 6 is completed.
KokkosP: Exiting profiling region: Realloc other view
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f881ec0 size: 161280
KokkosP: Allocate<Host> name: DualView::modified_flags pointer: 0x55fc9f881ec0 size: 8
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 7
KokkosP: Kokkos::View::initialization [DualView::modified_flags]
KokkosP: Execution of kernel 7 is completed.
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f882080 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 8
KokkosP: Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 8 is completed.
KokkosP: Entering profiling region: Resize same dualview
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f8a9800 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 9
KokkosP: Resize same dualview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 9 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 10
KokkosP: Resize same dualview
KokkosP:   Kokkos::ViewCopy-8D
KokkosP: Execution of kernel 10 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f882080 size: 161280
KokkosP: Exiting profiling region: Resize same dualview
KokkosP: Entering profiling region: Resize other dualview
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f8d0ec0 size: 181440
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 11
KokkosP: Resize other dualview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 11 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 12
KokkosP: Resize other dualview
KokkosP:   Kokkos::ViewCopy-8D
KokkosP: Execution of kernel 12 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f8a9800 size: 161280
KokkosP: Exiting profiling region: Resize other dualview
KokkosP: Entering profiling region: Realloc same dualview
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f8fd440 size: 181440
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 13
KokkosP: Realloc same dualview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 13 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f8d0ec0 size: 181440
KokkosP: Exiting profiling region: Realloc same dualview
KokkosP: Entering profiling region: Realloc other dualview
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f882080 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 14
KokkosP: Realloc other dualview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 14 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f8fd440 size: 181440
KokkosP: Exiting profiling region: Realloc other dualview
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f882080 size: 161280
KokkosP: Deallocate<Host> name: DualView::modified_flags pointer: 0x55fc9f881ec0 size: 8
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f882080 size: 6720
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 15
KokkosP: Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 15 is completed.
KokkosP: Entering profiling region: Resize same dynrankview
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f883bc0 size: 6720
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 16
KokkosP: Resize same dynrankview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 16 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 17
KokkosP: Resize same dynrankview
KokkosP:   Kokkos::DynRankViewRemap
KokkosP: Execution of kernel 17 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f882080 size: 6720
KokkosP: Exiting profiling region: Resize same dynrankview
KokkosP: Entering profiling region: Resize other dynrankview
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f8856c0 size: 7560
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 18
KokkosP: Resize other dynrankview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 18 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 19
KokkosP: Resize other dynrankview
KokkosP:   Kokkos::DynRankViewRemap
KokkosP: Execution of kernel 19 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f883bc0 size: 6720
KokkosP: Exiting profiling region: Resize other dynrankview
KokkosP: Entering profiling region: Realloc same dynrankview
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f8856c0 size: 7560
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f882080 size: 7560
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 20
KokkosP: Realloc same dynrankview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 20 is completed.
KokkosP: Exiting profiling region: Realloc same dynrankview
KokkosP: Entering profiling region: Realloc other dynrankview
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f882080 size: 7560
KokkosP: Allocate<Host> name: bla pointer: 0x55fc9f882080 size: 6720
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 21
KokkosP: Realloc other dynrankview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 21 is completed.
KokkosP: Exiting profiling region: Realloc other dynrankview
KokkosP: Deallocate<Host> name: bla pointer: 0x55fc9f882080 size: 6720
KokkosP: Kokkos library finalization called.
```
before this pull request and
```
KokkosP: Example Library Initialized (sequence is 0, version: 20210623)
KokkosP: Allocate<Host> name: bla pointer: 0x7f189b139040 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 0
KokkosP: Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 0 is completed.
KokkosP: Entering profiling region: Resize same view
KokkosP: Exiting profiling region: Resize same view
KokkosP: Entering profiling region: Resize other view
KokkosP: Allocate<Host> name: bla pointer: 0x7f189b0fe040 size: 181440
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 1
KokkosP: Resize other view
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 1 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 2
KokkosP: Resize other view
KokkosP:   Kokkos::ViewCopy-8D
KokkosP: Execution of kernel 2 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x7f189b139040 size: 161280
KokkosP: Exiting profiling region: Resize other view
KokkosP: Entering profiling region: Realloc same view
KokkosP: DeepCopy<Host,Host> DST(name: bla pointer: 0x7f189b0fe0c0) SRC(name: Scalar pointer 0x7fff1d4e9440) Size: 181440
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 3
KokkosP: Realloc same view
KokkosP:   Kokkos::ViewFill-1D
KokkosP: Execution of kernel 3 is completed.
KokkosP: Exiting profiling region: Realloc same view
KokkosP: Entering profiling region: Realloc other view
KokkosP: Deallocate<Host> name: bla pointer: 0x7f189b0fe040 size: 181440
KokkosP: Allocate<Host> name: bla pointer: 0x563561917f00 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 4
KokkosP: Realloc other view
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 4 is completed.
KokkosP: Exiting profiling region: Realloc other view
KokkosP: Deallocate<Host> name: bla pointer: 0x563561917f00 size: 161280
KokkosP: Allocate<Host> name: DualView::modified_flags pointer: 0x563561917f00 size: 8
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 5
KokkosP: Kokkos::View::initialization [DualView::modified_flags]
KokkosP: Execution of kernel 5 is completed.
KokkosP: Allocate<Host> name: bla pointer: 0x5635619180c0 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 6
KokkosP: Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 6 is completed.
KokkosP: Entering profiling region: Resize same dualview
KokkosP: Exiting profiling region: Resize same dualview
KokkosP: Entering profiling region: Resize other dualview
KokkosP: Allocate<Host> name: bla pointer: 0x56356193f840 size: 181440
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 7
KokkosP: Resize other dualview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 7 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 8
KokkosP: Resize other dualview
KokkosP:   Kokkos::ViewCopy-8D
KokkosP: Execution of kernel 8 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x5635619180c0 size: 161280
KokkosP: Exiting profiling region: Resize other dualview
KokkosP: Entering profiling region: Realloc same dualview
KokkosP: DeepCopy<Host,Host> DST(name: bla pointer: 0x56356193f8c0) SRC(name: Scalar pointer 0x7fff1d4e9400) Size: 181440
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 9
KokkosP: Realloc same dualview
KokkosP:   Kokkos::ViewFill-1D
KokkosP: Execution of kernel 9 is completed.
KokkosP: Exiting profiling region: Realloc same dualview
KokkosP: Entering profiling region: Realloc other dualview
KokkosP: Allocate<Host> name: bla pointer: 0x5635619180c0 size: 161280
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 10
KokkosP: Realloc other dualview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 10 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x56356193f840 size: 181440
KokkosP: Exiting profiling region: Realloc other dualview
KokkosP: Deallocate<Host> name: bla pointer: 0x5635619180c0 size: 161280
KokkosP: Deallocate<Host> name: DualView::modified_flags pointer: 0x563561917f00 size: 8
KokkosP: Allocate<Host> name: bla pointer: 0x5635619180c0 size: 6720
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 11
KokkosP: Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 11 is completed.
KokkosP: Entering profiling region: Resize same dynrankview
KokkosP: Exiting profiling region: Resize same dynrankview
KokkosP: Entering profiling region: Resize other dynrankvieww
KokkosP: Allocate<Host> name: bla pointer: 0x563561919c00 size: 7560
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 12
KokkosP: Resize other dynrankvieww
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 12 is completed.
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 13
KokkosP: Resize other dynrankvieww
KokkosP:   Kokkos::DynRankViewRemap
KokkosP: Execution of kernel 13 is completed.
KokkosP: Deallocate<Host> name: bla pointer: 0x5635619180c0 size: 6720
KokkosP: Exiting profiling region: Resize other dynrankvieww
KokkosP: Entering profiling region: Realloc same dynrankview
KokkosP: Executing parallel-for kernel on device 1 with unique execution identifier 14
KokkosP: Realloc same dynrankview
KokkosP:   Kokkos::DynRankViewFill
KokkosP: Execution of kernel 14 is completed.
KokkosP: Exiting profiling region: Realloc same dynrankview
KokkosP: Entering profiling region: Realloc other dynrankview
KokkosP: Deallocate<Host> name: bla pointer: 0x563561919c00 size: 7560
KokkosP: Allocate<Host> name: bla pointer: 0x5635619180c0 size: 6720
KokkosP: Executing parallel-for kernel on device 0 with unique execution identifier 15
KokkosP: Realloc other dynrankview
KokkosP:   Kokkos::View::initialization [bla]
KokkosP: Execution of kernel 15 is completed.
KokkosP: Exiting profiling region: Realloc other dynrankview
KokkosP: Deallocate<Host> name: bla pointer: 0x5635619180c0 size: 6720
KokkosP: Kokkos library finalization called.
```
after it. The diff basically shows that we still use an allocation when calling `resize` with the same extents previously (despite the attempt to already address this for `Kokkos::View`s; this worked correctly already if there a now static extents, though). As expected `realloc` calls `Kokkos::deep_copy` instead of allocating if the extents match.

File: core/src/Kokkos_CopyViews.hpp [hunk 1]
```diff
@@ -3009,80 +2958,29 @@ resize(const I& arg_prop, Kokkos::View<T, P...>& v,
...
+  const size_t new_extents[8] = {n0, n1, n2, n3, n4, n5, n6, n7};
+  bool sizeMismatch           = false;
+  for (unsigned int dim = 0; dim < v.rank_dynamic; ++dim)
+    if (new_extents[dim] != v.extent(dim)) {
+      sizeMismatch = true;
+      break;
+    }
+
+  if (sizeMismatch) {
+    view_type v_resized(view_alloc(v.label(), std::forward<const I>(arg_prop)),
```
Comment (dalg24) [inline core/src/Kokkos_CopyViews.hpp:2977]: Not sure what this forwarding is supposed to do achieve.  @nliber ?

**Review by dalg24 — COMMENTED** [review]
You forgot one `resize` overload
https://github.com/kokkos/kokkos/blob/73fa2270f33dc3836c8082b5353e5fe35bb17a8b/core/src/Kokkos_CopyViews.hpp#L2988-L2990

**Comment by masterleinad** [comment]
> You forgot one `resize` overload

I refrained from doing anything with this overload first since it requires comparing array layouts. I added comparison operators now but we should discuss if we want to have this in this pull request.

**Comment by stanmoore1** [comment]
The `DualView::resize` code looks much better than before.

File: core/src/Kokkos_CopyViews.hpp [hunk 2]
```diff
@@ -3095,11 +2993,13 @@ inline void resize(Kokkos::View<T, P...>& v,
   static_assert(Kokkos::ViewTraits<T, P...>::is_managed,
                 "Can only resize managed views");
 
-  view_type v_resized(v.label(), layout);
+  if (v.layout() != layout) {
```
Comment (crtrott) [inline core/src/Kokkos_CopyViews.hpp:2996]: Hm this is problematic since there are custom layouts in circulation.

**Review by crtrott — CHANGES_REQUESTED** [review]
Looks pretty good but we need to discuss the Layout changes.

File: core/src/Kokkos_CopyViews.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Kokkos_CopyViews.hpp:2996]: Yes, that was also my concern a little bit. I added it because @dalg24 asked for it and checked that it works as expected for at least `LayoutRight` but I would tend to exclude it from this pull request.

**Comment by masterleinad** [comment]
Retest this please.

File: containers/src/Kokkos_DualView.hpp [hunk 3]
```diff
@@ -875,8 +875,19 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
+      if (new_extents[dim] != h_view.extent(dim)) {
+        sizeMismatch = true;
+        break;
+      }
+
+    if (sizeMismatch) {
+      ::Kokkos::realloc(d_view, n0, n1, n2, n3, n4, n5, n6, n7);
+      h_view = create_mirror_view(d_view);
+    } else
+      ::Kokkos::deep_copy(d_view, typename t_dev::value_type{});
```
Comment (dalg24) [inline containers/src/Kokkos_DualView.hpp:890]: Why did you add this else-branch?
It was not present in the original code.

File: containers/src/Kokkos_DynRankView.hpp [hunk 4]
```diff
@@ -2063,10 +2073,21 @@ inline void realloc(DynRankView<T, P...>& v,
...
+
+  if (sizeMismatch) {
+    const std::string label = v.label();
 
-  v = drview_type();  // Deallocate first, if the only view to allocation
-  v = drview_type(label, n0, n1, n2, n3, n4, n5, n6, n7);
+    v = drview_type();  // Deallocate first, if the only view to allocation
+    v = drview_type(label, n0, n1, n2, n3, n4, n5, n6, n7);
+  } else
+    Kokkos::deep_copy(v, typename drview_type::value_type{});
```
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:2090]: Same question about the else-branch.

Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:2090]: Citing from https://github.com/kokkos/kokkos/pull/4207#issue-701805574:
> For resize we don't do anything if the extents already match but we call Kokkos::deep_copy for realloc to preserve the current behavior. This should of course change when we add the AllocateWithoutInitializing overloads.

I expect `DynRankView` to also zero initialize in the constructor and hence expect that `realloc` was also zero initializing before.

File: containers/src/Kokkos_DualView.hpp (hunk 3 above)
Comment (masterleinad) [reply containers/src/Kokkos_DualView.hpp:890]: Same here. The previous behavior was that `realloc` zero initializes the View. I am, of course, open to discuss not doing it. So far my thinking was to avoid omit it only when we pass `WithoutInitializing` (which is a later pull request someone else was assigned to IIRC).

Comment (dalg24) [reply containers/src/Kokkos_DualView.hpp:890]: Got it

**Comment by masterleinad** [comment]
@crtrott Wat should we do about the `array_layout` overload?

File: core/src/Kokkos_CopyViews.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/Kokkos_CopyViews.hpp:2978]: This was introduced in https://github.com/kokkos/kokkos/pull/2285 and I would rather address that in the `WithoutInitiializing` refactor @janciesko is supposed to do.

**Comment by masterleinad** [comment]
We decided to only try to compare array layouts for the built-in layouts (and I implemented that in the last commit).

File: containers/src/Kokkos_DualView.hpp [hunk 5]
```diff
@@ -875,8 +875,15 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
                const size_t n7 = KOKKOS_IMPL_CTOR_DEFAULT_ARG) {
-    ::Kokkos::realloc(d_view, n0, n1, n2, n3, n4, n5, n6, n7);
-    h_view = create_mirror_view(d_view);
+    const size_t new_extents[8] = {n0, n1, n2, n3, n4, n5, n6, n7};
+    const bool sizeMismatch =
+        Impl::size_mismatch(h_view, h_view.rank_dynamic, new_extents);
+
+    if (sizeMismatch) {
+      ::Kokkos::realloc(d_view, n0, n1, n2, n3, n4, n5, n6, n7);
+      h_view = create_mirror_view(d_view);
```
Comment (crtrott) [inline containers/src/Kokkos_DualView.hpp:884]: should this use the version which doesn't initialize?

File: containers/src/Kokkos_DualView.hpp (hunk 3 above)
Comment (crtrott) [reply containers/src/Kokkos_DualView.hpp:886]: ok

File: containers/src/Kokkos_DualView.hpp (hunk 5 above)
Comment (crtrott) [reply containers/src/Kokkos_DualView.hpp:884]: ah saw the comment below

**Review by crtrott — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/Kokkos_CopyViews.hpp [hunk 6]
```diff
@@ -2875,6 +2876,19 @@ inline void deep_copy(
 
 namespace Kokkos {
 
+namespace Impl {
+template <typename ViewType>
```
Comment (e10harvey) [inline core/src/Kokkos_CopyViews.hpp:2880]: Please add function documentation for `size_mismatch`

File: core/src/Kokkos_CopyViews.hpp [hunk 7]
```diff
@@ -2994,90 +2952,71 @@ resize(const I& arg_prop, Kokkos::View<T, P...>& v,
...
-  Kokkos::fence("Kokkos::resize(View)");
+  if (sizeMismatch) {
+    view_type v_resized(view_alloc(v.label(), std::forward<const I>(arg_prop)),
+                        n0, n1, n2, n3, n4, n5, n6, n7);
 
-  v = v_resized;
+    Kokkos::Impl::ViewRemap<view_type, view_type>(v_resized, v);
+    // This fence really ought to look for an execution space in
+    // arg_prop, and just fence that if there is one
+    Kokkos::fence("Kokkos::resize(View)");
```
Comment (e10harvey) [inline core/src/Kokkos_CopyViews.hpp:2972]: Same as above

File: core/src/Kokkos_CopyViews.hpp [hunk 8]
```diff
@@ -3112,10 +3050,16 @@ realloc(Kokkos::View<T, P...>& v,
...
 
-  v = view_type();  // Deallocate first, if the only view to allocation
-  v = view_type(label, n0, n1, n2, n3, n4, n5, n6, n7);
+  if (sizeMismatch) {
+    const std::string label = v.label();
+
+    v = view_type();  // Deallocate first, if the only view to allocation
+    v = view_type(label, n0, n1, n2, n3, n4, n5, n6, n7);
+  } else
+    Kokkos::deep_copy(v, typename view_type::value_type{});
```
Comment (e10harvey) [inline core/src/Kokkos_CopyViews.hpp:3062]: Is this writing zeros to `v`?

File: core/src/Kokkos_CopyViews.hpp [hunk 9]
```diff
@@ -2896,80 +2910,24 @@ resize(Kokkos::View<T, P...>& v, const size_t n0 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
...
+  const bool sizeMismatch = Impl::size_mismatch(v, v.rank_dynamic, new_extents);
 
-  Kokkos::Impl::ViewRemap<view_type, view_type>(v_resized, v);
-  Kokkos::fence("Kokkos::resize(View)");
+  if (sizeMismatch) {
+    view_type v_resized(v.label(), n0, n1, n2, n3, n4, n5, n6, n7);
 
-  v = v_resized;
+    Kokkos::Impl::ViewRemap<view_type, view_type>(v_resized, v);
+    Kokkos::fence("Kokkos::resize(View)");
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/Kokkos_CopyViews.hpp:2927]: or at least potentially asycn execution

File: core/src/Kokkos_CopyViews.hpp (hunk 8 above)
Comment (crtrott) [reply core/src/Kokkos_CopyViews.hpp:3062]: yes

File: core/src/Kokkos_Layout.hpp [hunk 10]
```diff
@@ -88,6 +88,16 @@ struct LayoutLeft {
...
+
+  friend bool operator==(const LayoutLeft& left, const LayoutLeft& right) {
+    for (unsigned int rank = 0; rank < ARRAY_LAYOUT_MAX_RANK; ++rank)
+      if (left.dimension[rank] != right.dimension[rank]) return false;
+    return true;
+  }
+
+  friend bool operator!=(const LayoutLeft& left, const LayoutLeft& right) {
+    return !(left == right);
+  }
```
[the remark this answers is not in the document]
Comment (e10harvey) [reply core/src/Kokkos_Layout.hpp:100]: OK. These 8 operators for `LayoutLeft`, `LayoutRight`, `LayoutStride`, `LayoutTiled` are candidates for refactoring into a base class `Layout` which Left, Right, Stride, and Tiled inherit from.

**Review by e10harvey — APPROVED** [review]

**Comment by ajpowelsnl** [comment]
@dalg24 , @crtrott , @masterleinad - can we close this issue?
